### ISSUE TITLE
Dependent styles fail during render rendering

### ISSUE THREAD

**Issue by rmzelle** [citation-style-language/styles#5274]
This is a rather long-standing issue IIRC, but dependent styles don't show previews. See e.g. https://github.com/citation-style-language/styles/pull/5271#issuecomment-778933173 with accompanying passing build https://github.com/citation-style-language/styles/runs/1900657050:

>Below are some sample citations generated based on your proposed changes:
dependent/vegetation-classification-and-survey.csl (rendering failure)

This may be especially confusing to outside contributors who think there might be a problem with the dependent style they're submitting. I guess we could either clarify the message (and e.g. skip dependents from previews altogether), or get dependent preview to work properly. Dependent styles have the option to overwrite the locale of their independent parent, so they don't necessary give the same output.

@adam3smith, @retorquere

**Comment by adam3smith** [comment]
Yeah, it has definitely caused occasional confusion. I don't think it's important to render dependent styles if that involves any serious effort. Simply not displaying the rendering attempt or a message like "No preview is generated for dependent styles" is perfect.

**Comment by retorquere** [comment]
I don't really know how the tests themselves work, and I'm trying to change as little as possible; I'm really not current enough in ruby and have never used `spec`. I wouldn't know where to begin to add substyle rendering.

**Comment by rmzelle** [comment]
> I don't really know how the tests themselves work, and I'm trying to change as little as possible; I'm really not current enough in ruby and have never used `spec`. I wouldn't know where to begin to add dependent style rendering.

I might make an attempt at this myself. I don't know how to fix rendering of dependents, but it would already be an improvement if the comment would just mention something like "dependent style, not rendered" instead of "rendering failure".

The existing logic for that seems pretty easy to modify for that (https://github.com/citation-style-language/Sheldon/blob/e115524cf1342613f694392985997c57c66db0b4/bin/sheldon#L358 just needs to have another `elsif` near the end to give a better message if rendering failed and the style was a dependent. The function already seems to have access to the style path (`style.path`), which presumably includes "dependent/" we can check for with a simple string match.

**Comment by retorquere** [comment]
> I might make an attempt at this myself. I don't know how to fix rendering of dependents, but it would already be an improvement if the comment would just mention something like "dependent style, not rendered" instead of "rendering failure".

That should be feasible to do. I can see the paths being offered.
 
> The existing logic for that seems pretty easy to modify for that (https://github.com/citation-style-language/Sheldon/blob/e115524cf1342613f694392985997c57c66db0b4/bin/sheldon#L358 just needs to have another `elsif` near the end to give a better message if rendering failed and the style was a dependent. The function already seems to have access to the style path (`style.path`), which presumably includes "dependent/" we can check for with a simple string match.

Yep, that will do it.

What would be needed to render these dependent styles?

**Comment by rmzelle** [comment]
> What would be needed to render these dependent styles?

You'd want to render the independent parent of the dependent style, while overriding the locale of that parent by that of the dependent (if specified). Not sure how to do that, though, and it's not a big issue if we don't get this working.

**Comment by retorquere** [comment]
My naive take on things is that it looks like the data from the dependent style will just replace the matching items in the independent style. Does that sound about right? I can take a stab at that (in my own repo this time).

**Comment by retorquere** [comment]
I've bunged [this](https://gist.github.com/retorquere/f6b299f00fe231a15765a512a4075059) together, but I don't know:

* whether field order matters within the `info` block -- right now, if I can't find a corresponding element, I just add it to the end of `info`
* whether I should remove all `category` elements if the dependent style has any -- right now, I leave `<category field="generic-base"/>` in place for example
* whether `rights` can just be replaced
* whether I should indeed just add `author`s and `contributor`s from the dependent style to the independent base
* what to do with `<link ... rel="template"/>`

**Comment by rmzelle** [comment]
> * whether field order matters within the `info` block -- right now, if I can't find a corresponding element, I just add it to the end of `info`

Sorry for the slow response! Nothing in `info` affects style rendering (although a CSL processor might expect some fields to be present like a style ID), so order doesn't matter.

**Comment by retorquere** [comment]
I don't understand how the tests are strung together -- I think this

```
CSL_TEST="dependent/zwr-das-deutsche-zahnarzteblatt.csl" bundle exec rspec spec
```

should do something, but I don't see anything happening.
